govcd/vm.go

```diff
@@ -852,6 +853,315 @@ func (vm *VM) SetGuestCustomizationSection(guestCustomizationSection *types.Gues
+			return nil, fmt.Errorf("could not validate if NIC %d uses routed network attached to edge gateway: %s",
+				nic.vmNicIndex, err)
+		}
+
+		// This nicIndex is not attached to routed network, move further
+		if IsNotFound(err) {
+			util.Logger.Printf("[TRACE] VM '%s' NIC with index %d is not attached to edge gateway routed network\n",
+				vm.VM.Name, nic.vmNicIndex)
+			continue
+		}
+
+		// Lookup edge gateway
+		edgeGateway, err := vdc.GetEdgeGatewayByName(edgeGatewayName, false)
+		if err != nil {
+			return nil, fmt.Errorf("could not lookup edge gateway for routed network on NIC %d: %s",
+				nic.vmNicIndex, err)
+		}
+
+		util.Logger.Printf("[TRACE] VM '%s' NIC with index %d is attached to edge gateway routed network\n",
+			vm.VM.Name, nic.vmNicIndex)
+		nicDhcpConfigs[index].routedNetworkEdgeGateway = edgeGateway
+	}
+
+	return nicDhcpConfigs, nil
+}
+
+// nicDhcpConfig is used to group data for carrying between multiple functions and optimizing on API
+// calls
+type nicDhcpConfig struct {
+	vmNicIndex               int
+	ip                       string
+	mac                      string
+	routedNetworkEdgeGateway *EdgeGateway
+}
+
+// nicDhcpConfigs is a slice of nicDhcpConfig
+type nicDhcpConfigs []nicDhcpConfig
+
+// getIpsFromNicDhcpConfigs extracts just IP addresses from nicDhcpConfigs
+func getIpsFromNicDhcpConfigs(nicConfigs []nicDhcpConfig) []string {
+	result := make([]string, len(nicConfigs))
+	for index, nicConfig := range nicConfigs {
+		result[index] = nicConfig.ip
+	}
+	return result
+}
+
+// allNicsHaveIps checks if all nicDhcpConfig in slice have not empty IP field
+func allNicsHaveIps(nicConfigs []nicDhcpConfig) bool {
+	allNicsHaveIps := true
+	for _, nicConfig := range nicConfigs {
+		if nicConfig.ip == "" {
+			allNicsHaveIps = false
+		}
+	}
+	return allNicsHaveIps
+}
+
+// WaitForDhcpIpByNicIndexes accepts a slice of NIC indexes in VM, tries to get these IPs up to
+// maxWaitSeconds and the returns:
+// * a list of IPs
+// * whether the function hit timeout (some IP values may be available after timeout)
+// * error
+// guaranteed to return all IP addresses on a certain timeout.
+//
+// This function checks a slice of nicIndexes and reuses all possible API calls. It may return a
+// partial result for IP addresses when the timeout is hit.
+//
+// Getting a DHCP address is complicated because vCD (in UI and in types.NetworkConnectionSection)
+// reports IP addresses when guest tools are present on a VM only. This function also attempts to
+// check if VM NICs are attached to routed network on edge gateway - then there is a chance that
+// built-in DHCP pools are used and active DHCP leases can be found.
+//
+// For this function to work - at least one the following must be true:
+// * VM has guest tools (vCD UI shows IP address). (Takes longer time)
+// * VM DHCP interface is connected to routed Org network and is using Edge Gateway DHCP. (Takes
+// less time, but is more constrained)
```

Comment:
The above two paragraphs are pretty amazing. I haven't checked yet, but hope similar level of detail is provided on the Terraform side :)

Reply:
Yes, it has pretty much copy/paste sentences :)